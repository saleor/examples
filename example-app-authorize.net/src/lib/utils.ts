import type { JSONValue } from "../types";
import { BaseError, UnknownError } from "@/errors";

export const tryJsonParse = (text: string | undefined) => {
  if (!text) {
    return undefined;
  }
  try {
    return JSON.parse(text) as JSONValue;
  } catch (e) {
    return text;
  }
};

export const tryIgnore = (fn: () => void) => {
  try {
    fn();
  } catch {
    // noop
  }
};

export const toStringOrEmpty = (value: unknown) => {
  if (typeof value === "string") return value;
  return "";
};

type PromiseToTupleResult<T> = [Error, null] | [null, Awaited<T>];
export const unpackPromise = async <T extends Promise<unknown>>(
  promise: T,
): Promise<PromiseToTupleResult<T>> => {
  try {
    const result = await promise;
    return [null, result];
  } catch (maybeError) {
    if (maybeError instanceof Error) {
      return [maybeError, null];
    }
    return [BaseError.normalize(maybeError, UnknownError), null];
  }
};

type ThrowableToTupleResult<T> = [Error, null] | [null, T];
export const unpackThrowable = <T>(throwable: () => T): ThrowableToTupleResult<T> => {
  try {
    const result = throwable();
    return [null, result];
  } catch (maybeError) {
    if (maybeError instanceof Error) {
      return [maybeError, null];
    }
    return [BaseError.normalize(maybeError, UnknownError), null];
  }
};

export const isNotNullish = <T>(val: T | null | undefined): val is T =>
  val !== undefined && val !== null;

export const isObject = (val: unknown): val is Record<string, unknown> =>
  typeof val === "object" && val !== null && !Array.isArray(val);

export const __do = <T>(fn: () => T): T => fn();
