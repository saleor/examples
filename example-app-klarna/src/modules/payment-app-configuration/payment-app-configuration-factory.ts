import { type Client } from "urql";
import { type MetadataEntry } from "@saleor/app-sdk/settings-manager";
import {
  createPrivateSettingsManager,
  createWebhookPrivateSettingsManager,
} from "../app-configuration/metadata-manager";
import { PaymentAppConfigurator } from "./payment-app-configuration";

export const getPaymentAppConfigurator = (client: Client, saleorApiUrl: string) => {
  return new PaymentAppConfigurator(createPrivateSettingsManager(client), saleorApiUrl);
};

export const getWebhookPaymentAppConfigurator = (
  data: { privateMetadata: readonly Readonly<MetadataEntry>[] },
  saleorApiUrl: string,
) => {
  return new PaymentAppConfigurator(
    createWebhookPrivateSettingsManager(data.privateMetadata as MetadataEntry[]),
    saleorApiUrl,
  );
};
