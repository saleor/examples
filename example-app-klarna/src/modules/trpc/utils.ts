import { type TRPCClientErrorLike } from "@trpc/client";
import {
  type DeepPartial,
  type FieldPath,
  type FieldValues,
  type UseFormSetError,
} from "react-hook-form";
import { type AppBridge } from "@saleor/app-sdk/app-bridge";
import { type AppRouter } from "./trpc-app-router";
import { BaseTrpcError, FieldError } from "@/errors";

interface HandlerInput {
  message?: string;
  title?: string;
  actionId: string;
  appBridge: AppBridge | undefined;
}

const getParsedError = <T extends TRPCClientErrorLike<AppRouter>>(error: T) => {
  if (error.data?.serialized) {
    return BaseTrpcError.parse(error.data.serialized);
  }
  return null;
};

export const getErrorHandler =
  (input: HandlerInput) =>
  <T extends TRPCClientErrorLike<AppRouter>>(error: T) => {
    if (input.appBridge) {
      const parsedError = getParsedError(error);
      void input.appBridge.dispatch({
        type: "notification",
        payload: {
          title: input.title || "Request failed",
          text: input.message || parsedError?.message || error.message,
          apiMessage: error.shape ? JSON.stringify(error.shape ?? {}, null, 2) : undefined,
          actionId: input.actionId,
          status: "error",
        },
      });
    }
  };

interface FieldHandlerInput<TFieldValues extends FieldValues> extends HandlerInput {
  fieldName: FieldPath<TFieldValues> | `root.${string}` | "root";
  formFields: FieldPath<TFieldValues>[];
  setError: UseFormSetError<TFieldValues>;
}

export const getFormFields = <TFieldValues extends FieldValues>(
  defaultValues: Readonly<DeepPartial<TFieldValues>> | undefined,
): FieldPath<TFieldValues>[] => {
  if (!defaultValues) {
    return [];
  }
  return Object.keys(defaultValues) as FieldPath<TFieldValues>[];
};

const isMatchingField = <TFieldValues extends FieldValues>(
  fieldName: string,
  formFields: FieldPath<TFieldValues>[],
): fieldName is FieldPath<TFieldValues> => {
  return formFields.some((field) => field === fieldName);
};

export const getFieldErrorHandler =
  <TFieldValues extends FieldValues>(input: FieldHandlerInput<TFieldValues>) =>
  <T extends TRPCClientErrorLike<AppRouter>>(error: T) => {
    getErrorHandler(input)(error);
    const parsedError = BaseTrpcError.parse(error.data?.serialized);

    if (
      parsedError instanceof FieldError &&
      isMatchingField(parsedError.fieldName, input.formFields)
    ) {
      input.setError(parsedError.fieldName, { message: parsedError.message });
    } else {
      input.setError(input.fieldName, {
        message: input.message,
      });
    }
  };
