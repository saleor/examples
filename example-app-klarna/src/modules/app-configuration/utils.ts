import { isNotNullish, toStringOrEmpty } from "../../lib/utils";

export const OBFUSCATION_DOTS = "••••";

export const obfuscateValue = (value: string) => {
  const unbofuscatedLength = Math.min(4, value.length - 4);

  if (unbofuscatedLength <= 0) {
    return OBFUSCATION_DOTS;
  }

  const visibleValue = value.slice(-unbofuscatedLength);
  return `${OBFUSCATION_DOTS}${visibleValue}`;
};

export const deobfuscateValues = (values: Record<string, unknown>) => {
  const entries = Object.entries(values).map(
    ([key, value]) =>
      [key, toStringOrEmpty(value).includes(OBFUSCATION_DOTS) ? null : value] as [string, unknown],
  );
  return Object.fromEntries(entries);
};

export const filterConfigValues = <T extends Record<string, unknown>>(values: T) => {
  const entries = Object.entries(values).filter(
    ([_, value]) => value !== null && value !== undefined,
  );
  return Object.fromEntries(entries);
};

export const obfuscateConfig = <T extends {}>(config: T): T => {
  const entries = Object.entries(config).map(([key, value]) => [
    key,
    isNotNullish(value) ? obfuscateValue(toStringOrEmpty(value)) : value,
  ]);

  return Object.fromEntries(entries) as T;
};
