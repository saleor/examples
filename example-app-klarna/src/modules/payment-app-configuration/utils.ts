import { obfuscateConfig } from "../app-configuration/utils";
import {
  type PaymentAppConfigEntry,
  type PaymentAppEncryptedConfig,
  type PaymentAppUserVisibleConfigEntry,
  paymentAppUserVisibleConfigEntrySchema,
} from "./config-entry";

export const obfuscateConfigEntry = (
  entry: PaymentAppConfigEntry,
): PaymentAppUserVisibleConfigEntry => {
  const { apiUrl, username, password, configurationName, configurationId } = entry;

  const configValuesToObfuscate = {
    password,
  } satisfies PaymentAppEncryptedConfig;

  return paymentAppUserVisibleConfigEntrySchema.parse({
    apiUrl,
    username,
    configurationId,
    configurationName,
    ...obfuscateConfig(configValuesToObfuscate),
  } satisfies PaymentAppUserVisibleConfigEntry);
};
