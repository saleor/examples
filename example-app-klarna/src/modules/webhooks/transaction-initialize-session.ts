import { z } from "zod";
import { getNormalizedLocale } from "@/backend-lib/api-route-utils";
import { KlarnaHttpClientError } from "@/errors";
import { env } from "@/lib/env.mjs";
import { invariant } from "@/lib/invariant";
import { createLogger } from "@/lib/logger";
import { obfuscateConfig } from "@/modules/app-configuration/utils";
import { getKlarnaIntegerAmountFromSaleor } from "@/modules/klarna/currencies";
import {
  getKlarnaApiClient,
  getLineItems,
  prepareRequestAddress,
  type KlarnaMetadata,
} from "@/modules/klarna/klarna-api";
import { paymentAppFullyConfiguredEntrySchema } from "@/modules/payment-app-configuration/config-entry";
import { getConfigurationForChannel } from "@/modules/payment-app-configuration/payment-app-configuration";
import { getWebhookPaymentAppConfigurator } from "@/modules/payment-app-configuration/payment-app-configuration-factory";
import { type TransactionInitializeSessionResponse } from "@/schemas/TransactionInitializeSession/TransactionInitializeSessionResponse.mjs";
import {
  TransactionFlowStrategyEnum,
  type TransactionInitializeSessionEventFragment,
} from "generated/graphql";
import { type components as hppComponents } from "generated/klarna-hpp";
import { type components as paymentsComponents } from "generated/klarna-payments";

const transactionInitializePayloadData = z.object({
  merchantUrls: z.object({
    success: z.string().url(),
    cancel: z.string().url().optional(),
    back: z.string().url().optional(),
    failure: z.string().url().optional(),
    error: z.string().url().optional(),
  }),
});

export const TransactionInitializeSessionWebhookHandler = async (
  event: TransactionInitializeSessionEventFragment,
  { saleorApiUrl, baseUrl }: { saleorApiUrl: string; baseUrl: string },
): Promise<TransactionInitializeSessionResponse> => {
  const appBaseUrl = env.APP_API_BASE_URL ?? baseUrl;

  const logger = createLogger(
    { saleorApiUrl },
    { msgPrefix: "[TransactionInitializeSessionWebhookHandler] " },
  );
  const { transaction, action, sourceObject, merchantReference, issuingPrincipal } = event;
  const { id, __typename, channel } = sourceObject;
  const logData = {
    transaction,
    action,
    sourceObject: { id, channel, __typename },
    merchantReference,
    issuingPrincipal,
  };
  logger.debug(logData, "Received event");

  const app = event.recipient;
  invariant(app, "Missing event.recipient!");
  invariant(event.data, "Missing data");

  const { merchantUrls } = transactionInitializePayloadData.parse(event.data);

  const { privateMetadata } = app;

  const configurator = getWebhookPaymentAppConfigurator({ privateMetadata }, saleorApiUrl);
  const appConfig = await configurator.getConfig();
  const klarnaConfig = paymentAppFullyConfiguredEntrySchema.parse(
    getConfigurationForChannel(appConfig, event.sourceObject.channel.id),
  );

  const klarnaClient = getKlarnaApiClient({
    klarnaApiUrl: klarnaConfig.apiUrl,
    username: klarnaConfig.username,
    password: klarnaConfig.password,
  });

  const createKlarnaSession = klarnaClient.path("/payments/v1/sessions").method("post").create();

  const createHppSession = klarnaClient.path("/hpp/v1/sessions").method("post").create();

  const locale = getNormalizedLocale(event);

  const country = event.sourceObject.billingAddress?.country.code;
  invariant(country, "Missing country code");

  const transactionId = event.transaction.id;
  const channelId = event.sourceObject.channel.id;

  const metadata: KlarnaMetadata = {
    transactionId,
    channelId,
    ...(event.sourceObject.__typename === "Checkout" && { checkoutId: event.sourceObject.id }),
    ...(event.sourceObject.__typename === "Order" && { orderId: event.sourceObject.id }),
  };

  const orderLines = getLineItems(event.sourceObject);
  const orderTaxAmount = orderLines.reduce((acc, line) => acc + (line.total_tax_amount ?? 0), 0);

  const authorizationCallbackUrl = new URL(appBaseUrl);
  authorizationCallbackUrl.pathname = "/api/webhooks/klarna/authorization";
  authorizationCallbackUrl.searchParams.set("transactionId", transactionId);
  authorizationCallbackUrl.searchParams.set("channelId", channelId);
  authorizationCallbackUrl.searchParams.set("saleorApiUrl", saleorApiUrl);

  const email = sourceObject.userEmail;
  const createKlarnaSessionPayload: paymentsComponents["schemas"]["session_create"] = {
    locale: locale.split("_")[0],
    purchase_country: country,
    purchase_currency: event.action.currency,
    billing_address: prepareRequestAddress(sourceObject.billingAddress, email),
    shipping_address: prepareRequestAddress(sourceObject.shippingAddress, email),
    order_amount: getKlarnaIntegerAmountFromSaleor(event.action.amount, event.action.currency),
    order_tax_amount: orderTaxAmount,
    order_lines: orderLines,
    intent: "buy",
    merchant_reference1: event.transaction.id,
    merchant_reference2: event.sourceObject.id,
    merchant_data: JSON.stringify(metadata),
    merchant_urls: {
      authorization: authorizationCallbackUrl.toString(),
    },
  };

  logger.info(authorizationCallbackUrl.toString());
  logger.debug({ ...obfuscateConfig(createKlarnaSessionPayload) }, "createKlarnaSession payload");

  const klarnaSession = await createKlarnaSession(createKlarnaSessionPayload);

  logger.debug({ ...obfuscateConfig(klarnaSession) }, "createKlarnaSession result");

  if (!klarnaSession.ok) {
    throw new KlarnaHttpClientError(klarnaSession.statusText, { errors: [klarnaSession.data] });
  }

  const baseSuccessUrl = new URL(merchantUrls.success);
  baseSuccessUrl.searchParams.append("authorization_token", "{{authorization_token}}");
  baseSuccessUrl.searchParams.append("transaction_id", transactionId);
  // dont encode the search params because {{authorization_token}} is a placeholder
  baseSuccessUrl.search = decodeURIComponent(baseSuccessUrl.search);

  const successUrl = baseSuccessUrl.toString();

  const createHppSessionPayload: hppComponents["schemas"]["SessionCreationRequestV1"] = {
    payment_session_url:
      klarnaConfig.apiUrl + "/payments/v1/sessions/" + klarnaSession.data.session_id,
    merchant_urls: {
      success: successUrl,
    },
  };

  const klarnaHpp = await createHppSession(createHppSessionPayload);

  if (!klarnaHpp.ok) {
    throw new KlarnaHttpClientError(klarnaHpp.statusText, { errors: [klarnaHpp.data] });
  }

  invariant(klarnaHpp.data.redirect_url, "Missing redirect_url in klarnaHpp response");

  const transactionInitializeSessionResponse: TransactionInitializeSessionResponse = {
    data: {
      klarnaHppResponse: {
        redirectUrl: klarnaHpp.data.redirect_url,
      },
    },
    pspReference: klarnaHpp.data.session_id,
    result:
      event.action.actionType === TransactionFlowStrategyEnum.Authorization
        ? "AUTHORIZATION_ACTION_REQUIRED"
        : "CHARGE_ACTION_REQUIRED",
    actions: [],
    amount: action.amount,
    message: "",
    externalUrl: klarnaHpp.data.session_url,
  };
  return transactionInitializeSessionResponse;
};
