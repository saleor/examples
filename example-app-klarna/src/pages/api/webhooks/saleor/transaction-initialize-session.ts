import { SaleorSyncWebhook } from "@saleor/app-sdk/handlers/next";
import { type PageConfig } from "next";
import { uuidv7 } from "uuidv7";
import { saleorApp } from "@/saleor-app";
import {
  UntypedTransactionInitializeSessionDocument,
  type TransactionInitializeSessionEventFragment,
  TransactionFlowStrategyEnum,
  TransactionEventTypeEnum,
} from "generated/graphql";
import { TransactionInitializeSessionWebhookHandler } from "@/modules/webhooks/transaction-initialize-session";
import { getSyncWebhookHandler } from "@/backend-lib/api-route-utils";
import ValidateTransactionInitializeSessionResponse from "@/schemas/TransactionInitializeSession/TransactionInitializeSessionResponse.mjs";

export const config: PageConfig = {
  api: {
    bodyParser: false,
  },
};

export const transactionInitializeSessionSyncWebhook =
  new SaleorSyncWebhook<TransactionInitializeSessionEventFragment>({
    name: "TransactionInitializeSession",
    apl: saleorApp.apl,
    event: "TRANSACTION_INITIALIZE_SESSION",
    query: UntypedTransactionInitializeSessionDocument,
    webhookPath: "/api/webhooks/saleor/transaction-initialize-session",
  });

export default transactionInitializeSessionSyncWebhook.createHandler(
  getSyncWebhookHandler(
    "transactionInitializeSessionSyncWebhook",
    TransactionInitializeSessionWebhookHandler,
    ValidateTransactionInitializeSessionResponse,
    (payload, errorResponse) => {
      return {
        amount: 0,
        result:
          payload.action.actionType === TransactionFlowStrategyEnum.Authorization
            ? TransactionEventTypeEnum.AuthorizationFailure
            : TransactionEventTypeEnum.ChargeFailure,
        message: errorResponse.message,
        data: { errors: errorResponse.errors, klarnaHppResponse: {} },
        // @todo consider making pspReference optional https://github.com/saleor/saleor/issues/12490
        pspReference: uuidv7(),
      } as const;
    },
  ),
);
