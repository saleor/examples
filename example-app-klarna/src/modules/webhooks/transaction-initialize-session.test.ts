import { describe, it, expect } from "vitest";

import { TransactionInitializeSessionWebhookHandler } from "./transaction-initialize-session";
import {
  createMockTransactionInitializeSessionEvent,
  createMockTransactionInitializeSessionSourceObjectCheckout,
  createMockTransactionInitializeSessionSourceObjectOrder,
} from "./__tests__/utils";
import { setupRecording } from "@/__tests__/polly";
import { testEnv } from "@/__tests__/test-env.mjs";

import { LanguageCodeEnum, TransactionFlowStrategyEnum } from "generated/graphql";
import { filledFakeMatadataConfig } from "@/modules/payment-app-configuration/__tests__/utils";
import { getFakePaymentAppConfigurator } from "@/modules/payment-app-configuration/__tests__/payment-app-configuration-factory";

describe("TransactionInitializeSessionWebhookHandler", () => {
  setupRecording({});
  describe.each([
    "https://api.playground.klarna.com/",
    // 429 Too Many Requests
    // "https://api-na.playground.klarna.com/",
    // "https://api-oc.playground.klarna.com/",
  ])("%p", (apiUrl) => {
    describe.each([
      {
        name: "Checkout",
        getSourceObject: createMockTransactionInitializeSessionSourceObjectCheckout,
      },
      { name: "Order", getSourceObject: createMockTransactionInitializeSessionSourceObjectOrder },
    ])("$name", ({ getSourceObject }) => {
      it("should work", async () => {
        const configurator = getFakePaymentAppConfigurator(
          {
            ...filledFakeMatadataConfig,
            configurations: [
              {
                ...filledFakeMatadataConfig.configurations[0],
                apiUrl,
              },
            ],
          },
          testEnv.TEST_SALEOR_API_URL,
        );
        const privateMetadata = await configurator.getRawConfig();

        const event = await createMockTransactionInitializeSessionEvent({
          data: {
            merchantUrls: {
              success: "https://example.com/success",
              cancel: "https://example.com/cancel",
              back: "https://example.com/back",
              failure: "https://example.com/failure",
              error: "https://example.com/error",
            },
          },
          action: {
            amount: 99.99 + 123.0,
            currency: "SEK",
            actionType: TransactionFlowStrategyEnum.Authorization,
          },
          issuingPrincipal: null,
          sourceObject: getSourceObject({
            languageCode: LanguageCodeEnum.En,
            total: {
              gross: {
                amount: 99.99 + 123.0,
                currency: "SEK",
              },
            },
            billingAddress: {
              country: {
                code: "SE",
              },
            },
          }),
          recipient: { privateMetadata },
        });
        const initializeResult = await TransactionInitializeSessionWebhookHandler(event, {
          saleorApiUrl: testEnv.TEST_SALEOR_API_URL,
          baseUrl: testEnv.APP_API_BASE_URL,
        });
        expect(initializeResult.data?.klarnaHppResponse).toMatchObject({
          redirectUrl: expect.any(String),
        });
        expect(initializeResult.result).toBe("AUTHORIZATION_ACTION_REQUIRED");
        expect(initializeResult.data).toMatchSnapshot();
      });

      it.todo("should work with partial payment", async () => {
        const configurator = getFakePaymentAppConfigurator(
          {
            ...filledFakeMatadataConfig,
            configurations: [
              {
                ...filledFakeMatadataConfig.configurations[0],
                apiUrl,
              },
            ],
          },
          testEnv.TEST_SALEOR_API_URL,
        );
        const privateMetadata = await configurator.getRawConfig();

        // total is 30
        // shipping is 10
        // line is 2 * 10
        // but we request payment for just 20
        const event = await createMockTransactionInitializeSessionEvent({
          data: {
            merchantUrls: {
              success: "https://example.com/success",
              cancel: "https://example.com/cancel",
              back: "https://example.com/back",
              failure: "https://example.com/failure",
              error: "https://example.com/error",
            },
          },
          action: {
            amount: 20,
            currency: "SEK",
            actionType: TransactionFlowStrategyEnum.Authorization,
          },
          issuingPrincipal: null,
          sourceObject: getSourceObject({
            languageCode: LanguageCodeEnum.En,
            shippingPrice: {
              gross: {
                amount: 10,
                currency: "SEK",
              },
              net: {
                amount: 8.13,
                currency: "SEK",
              },
              tax: {
                amount: 1.87,
                currency: "SEK",
              },
            },
            total: {
              gross: {
                amount: 30,
                currency: "SEK",
              },
            },
            lines: [
              {
                totalPrice: {
                  gross: {
                    amount: 20,
                    currency: "SEK",
                  },
                  net: {
                    amount: 16.26,
                    currency: "SEK",
                  },
                  tax: {
                    amount: 3.74,
                    currency: "SEK",
                  },
                },
                unitPrice: {
                  gross: {
                    amount: 10,
                    currency: "SEK",
                  },
                  net: {
                    amount: 8.13,
                    currency: "SEK",
                  },
                  tax: {
                    amount: 1.87,
                    currency: "SEK",
                  },
                },
                quantity: 2,
              },
            ],
            billingAddress: {
              country: {
                code: "SE",
              },
            },
          }),
          recipient: { privateMetadata },
        });
        const initializeResult = await TransactionInitializeSessionWebhookHandler(event, {
          saleorApiUrl: testEnv.TEST_SALEOR_API_URL,
          baseUrl: testEnv.APP_API_BASE_URL,
        });
        expect(initializeResult.data?.klarnaHppResponse).toMatchObject({
          redirectUrl: expect.any(String),
        });
        expect(initializeResult.result).toBe("AUTHORIZATION_ACTION_REQUIRED");
        expect(initializeResult.data).toMatchSnapshot();
      });
    });
  });
});
