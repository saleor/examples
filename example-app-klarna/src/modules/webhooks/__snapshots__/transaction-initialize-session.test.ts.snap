// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionInitializeSessionWebhookHandler > %p https://api.playground.klarna.com/ > 'Checkout' > should work 1`] = `
{
  "klarnaSessionResponse": {
    "client_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjgyMzA1ZWJjLWI4MTEtMzYzNy1hYTRjLTY2ZWNhMTg3NGYzZCJ9.eyJzZXNzaW9uX2lkIjoiOGViOGUyNmUtMDFhNy01NmVlLWI4NDctM2FkZDI0MzM0NWY5IiwiYmFzZV91cmwiOiJodHRwczovL2pzLnBsYXlncm91bmQua2xhcm5hLmNvbS9ldS9rcCIsImRlc2lnbiI6ImtsYXJuYSIsImxhbmd1YWdlIjoiZW4iLCJwdXJjaGFzZV9jb3VudHJ5IjoiU0UiLCJlbnZpcm9ubWVudCI6InBsYXlncm91bmQiLCJtZXJjaGFudF9uYW1lIjoiWW91ciBidXNpbmVzcyBuYW1lIiwic2Vzc2lvbl90eXBlIjoiUEFZTUVOVFMiLCJjbGllbnRfZXZlbnRfYmFzZV91cmwiOiJodHRwczovL2V1LnBsYXlncm91bmQua2xhcm5hZXZ0LmNvbSIsInNjaGVtZSI6dHJ1ZSwiZXhwZXJpbWVudHMiOlt7Im5hbWUiOiJrcC1jbGllbnQtb25lLXB1cmNoYXNlLWZsb3ciLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3BjLTFrLXNlcnZpY2UiLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3AtY2xpZW50LXV0b3BpYS1mbG93IiwidmFyaWF0ZSI6InZhcmlhdGUtMSJ9LHsibmFtZSI6ImtwYy1QU0VMLTMwOTkiLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3AtY2xpZW50LXV0b3BpYS1zdGF0aWMtd2lkZ2V0IiwidmFyaWF0ZSI6ImluZGV4IiwicGFyYW1ldGVycyI6eyJkeW5hbWljIjoidHJ1ZSJ9fSx7Im5hbWUiOiJpbi1hcHAtc2RrLW5ldy1pbnRlcm5hbC1icm93c2VyIiwicGFyYW1ldGVycyI6eyJ2YXJpYXRlX2lkIjoibmV3LWludGVybmFsLWJyb3dzZXItZW5hYmxlIn19LHsibmFtZSI6ImtwLWNsaWVudC11dG9waWEtc2RrLWZsb3ciLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3AtY2xpZW50LXV0b3BpYS13ZWJ2aWV3LWZsb3ciLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoiaW4tYXBwLXNkay1jYXJkLXNjYW5uaW5nIiwicGFyYW1ldGVycyI6eyJ2YXJpYXRlX2lkIjoiY2FyZC1zY2FubmluZy1lbmFibGUifX1dLCJyZWdpb24iOiJldSIsIm9yZGVyX2Ftb3VudCI6MjIyOTksIm9mZmVyaW5nX29wdHMiOjIsIm9vIjoiZDIiLCJ2ZXJzaW9uIjoidjEuMTAuMC0xNTkwLWczZWJjMzkwNyJ9.VlOMSkDst-VQR7UFSOStK3MvLBPsNbgaLxVimuayj1b1wGrygbXaozyit48nld8VCzAvowD7Hz8WioOu5irT5PJhmpbopOT0sR0N4zaNlwrice8a2ZX42xEy72scepW1lGNPP3ssIZ1jjxncOjNgm3UTV38Yn4rxmoLJJWgxSxAaKxZNm1ExLjkfVd84Q2oblk3WZ5ma6merDrLLt6EjvkvFQ4ZYxHbns_WmuVxDNXtK0kBPFoTX2-zbl22al2Hl5vK8kvjKJm6-KdirGE68qnI2tlOtOaIyNZartGvXINpaju8NLFAQH_YTPjCbgZHYnXdnmkrtc_4y3WqPwqP4zw",
    "payment_method_categories": [
      {
        "asset_urls": {
          "descriptive": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
          "standard": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
        },
        "identifier": "pay_later",
        "name": "Buy now, pay later",
      },
      {
        "asset_urls": {
          "descriptive": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
          "standard": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
        },
        "identifier": "pay_over_time",
        "name": "Monthly financing",
      },
      {
        "asset_urls": {
          "descriptive": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
          "standard": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
        },
        "identifier": "pay_now",
        "name": "Pay now",
      },
    ],
    "session_id": "8eb8e26e-01a7-56ee-b847-3add243345f9",
  },
}
`;

exports[`TransactionInitializeSessionWebhookHandler > %p https://api.playground.klarna.com/ > 'Order' > should work 1`] = `
{
  "klarnaSessionResponse": {
    "client_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjgyMzA1ZWJjLWI4MTEtMzYzNy1hYTRjLTY2ZWNhMTg3NGYzZCJ9.eyJzZXNzaW9uX2lkIjoiOTViNmMyNjEtOWExYS01OGQ3LWE3N2YtYzM5NDU0MjdjNTFiIiwiYmFzZV91cmwiOiJodHRwczovL2pzLnBsYXlncm91bmQua2xhcm5hLmNvbS9ldS9rcCIsImRlc2lnbiI6ImtsYXJuYSIsImxhbmd1YWdlIjoiZW4iLCJwdXJjaGFzZV9jb3VudHJ5IjoiU0UiLCJlbnZpcm9ubWVudCI6InBsYXlncm91bmQiLCJtZXJjaGFudF9uYW1lIjoiWW91ciBidXNpbmVzcyBuYW1lIiwic2Vzc2lvbl90eXBlIjoiUEFZTUVOVFMiLCJjbGllbnRfZXZlbnRfYmFzZV91cmwiOiJodHRwczovL2V1LnBsYXlncm91bmQua2xhcm5hZXZ0LmNvbSIsInNjaGVtZSI6dHJ1ZSwiZXhwZXJpbWVudHMiOlt7Im5hbWUiOiJrcC1jbGllbnQtb25lLXB1cmNoYXNlLWZsb3ciLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3BjLTFrLXNlcnZpY2UiLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3AtY2xpZW50LXV0b3BpYS1mbG93IiwidmFyaWF0ZSI6InZhcmlhdGUtMSJ9LHsibmFtZSI6ImtwYy1QU0VMLTMwOTkiLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3AtY2xpZW50LXV0b3BpYS1zdGF0aWMtd2lkZ2V0IiwidmFyaWF0ZSI6ImluZGV4IiwicGFyYW1ldGVycyI6eyJkeW5hbWljIjoidHJ1ZSJ9fSx7Im5hbWUiOiJpbi1hcHAtc2RrLW5ldy1pbnRlcm5hbC1icm93c2VyIiwicGFyYW1ldGVycyI6eyJ2YXJpYXRlX2lkIjoibmV3LWludGVybmFsLWJyb3dzZXItZW5hYmxlIn19LHsibmFtZSI6ImtwLWNsaWVudC11dG9waWEtc2RrLWZsb3ciLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoia3AtY2xpZW50LXV0b3BpYS13ZWJ2aWV3LWZsb3ciLCJ2YXJpYXRlIjoidmFyaWF0ZS0xIn0seyJuYW1lIjoiaW4tYXBwLXNkay1jYXJkLXNjYW5uaW5nIiwicGFyYW1ldGVycyI6eyJ2YXJpYXRlX2lkIjoiY2FyZC1zY2FubmluZy1lbmFibGUifX1dLCJyZWdpb24iOiJldSIsIm9yZGVyX2Ftb3VudCI6MjIyOTksIm9mZmVyaW5nX29wdHMiOjIsIm9vIjoiZDIiLCJ2ZXJzaW9uIjoidjEuMTAuMC0xNTkwLWczZWJjMzkwNyJ9.l6L0UWShbJ-EG9TSg_DFwM9fbB1wlbK8fxVVx-csE5fjvUOD2YvTmxnzCMs3JPVoJvNHdzj6DFZYiPoLGMWtJB0_Vf2SWWrnqWx_MpRWGvRfpWyFqvWtcL-W35ensdR_r1SBGJoj15COT9I0EM0qZspTsZM-teJ0ZkoyaO2amP6AtHOB8QtTDKPYcWanYmlL-DquBXP23Hm57KiM2HmPS-MSPNHrZm1yk2aJsZ_kFZ_ONadHs9apysXnhnPJ-LYOxMnwpUFpW22GNIzrAA7qGlYanB6fUS51g3gO8gwpn04OZBdlxtZjN6FPqkba5_AFaqmuH1zESGFx75mKR8EfAw",
    "payment_method_categories": [
      {
        "asset_urls": {
          "descriptive": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
          "standard": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
        },
        "identifier": "pay_later",
        "name": "Buy now, pay later",
      },
      {
        "asset_urls": {
          "descriptive": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
          "standard": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
        },
        "identifier": "pay_over_time",
        "name": "Monthly financing",
      },
      {
        "asset_urls": {
          "descriptive": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
          "standard": "https://x.klarnacdn.net/payment-method/assets/badges/generic/klarna.svg",
        },
        "identifier": "pay_now",
        "name": "Pay now",
      },
    ],
    "session_id": "95b6c261-9a1a-58d7-a77f-c3945427c51b",
  },
}
`;
